peerconnection.go

```diff
@@ -567,43 +562,63 @@ func (pc *PeerConnection) CreateAnswer(options *AnswerOptions) (SessionDescripti
 	d := sdp.NewJSEPSessionDescription(useIdentity)
 	pc.addFingerprint(d)
 
-	bundleValue := "BUNDLE"
-	for _, remoteMedia := range pc.RemoteDescription().parsed.MediaDescriptions {
-		// TODO @trivigy better SDP parser
-		var peerDirection RTPTransceiverDirection
-		midValue := ""
-		for _, a := range remoteMedia.Attributes {
+	getDirection := func(media *sdp.MediaDescription) RTPTransceiverDirection {
+		for _, a := range media.Attributes {
+			if direction := NewRTPTransceiverDirection(a.Key); direction != RTPTransceiverDirection(Unknown) {
+				return direction
+			}
+		}
+		return RTPTransceiverDirection(Unknown)
+	}
+
+	localTransceivers := append([]*RTPTransceiver{}, pc.GetTransceivers()...)
+	satisfyPeerMedia := func(kind RTPCodecType, direction RTPTransceiverDirection) *RTPTransceiver {
+		for i := range localTransceivers {
+			t := localTransceivers[i]
+
 			switch {
-			case strings.HasPrefix(*a.String(), "mid"):
-				midValue = (*a.String())[len("mid:"):]
-			case strings.HasPrefix(*a.String(), "sendrecv"):
-				peerDirection = RTPTransceiverDirectionSendrecv
-			case strings.HasPrefix(*a.String(), "sendonly"):
-				peerDirection = RTPTransceiverDirectionSendonly
-			case strings.HasPrefix(*a.String(), "recvonly"):
-				peerDirection = RTPTransceiverDirectionRecvonly
+			case t.kind != kind:
+				continue
+			case direction == RTPTransceiverDirectionSendrecv && t.Direction != RTPTransceiverDirectionSendrecv:
+				continue
+			case direction != RTPTransceiverDirectionSendrecv && direction == t.Direction:
+				continue
+			case direction == RTPTransceiverDirectionInactive:
```

Comment:
My application code is send-only (just calls AddTrack()), so I had to add the following statement in order to get this to work with chrome as offerer (direction.recvOnly):
```go
case direction == RTPTransceiverDirectionRecvonly && t.Direction == RTPTransceiverDirectionSendonly:
        fallthrough
```

Reply:
I am feeling dense, but which case here is causing a `continue` when remote is `recvonly` and we have a local `sendonly`? Reading I think it should fail all cases (and should return a local `Transceiver`)

I will add a test though, and reproduce locally?

Reply:
Sorry! I should have pasted in patch form; easier to see:

```diff
--- a/vendor/github.com/pions/webrtc/peerconnection.go
+++ b/vendor/github.com/pions/webrtc/peerconnection.go
@@ -582,6 +582,8 @@ func (pc *PeerConnection) CreateAnswer(options *AnswerOptions) (SessionDescripti
                                continue
                        case direction != RTPTransceiverDirectionSendrecv && direction == t.Direction:
                                continue
+                       case direction == RTPTransceiverDirectionRecvonly && t.Direction == RTPTransceiverDirectionSendonly:
+                               fallthrough
                        case direction == RTPTransceiverDirectionInactive:
                                continue
                        }```

Hmm. Did you update your PR? The last case (inactive) didn't have a continue when I looked at this earlier. I thought the logic was that if the offer's direction was inactive, it would satisfy that media entry with a localTransceiver. 

In any case, I'm probably not understanding the overall intent here. My patch was attempt to allow a localTransceiver with direction set to sendOnly to satisfy a remote media entry set to recvOnly. This worked in my testing.

Reply:
That was my mistake! I messed up the `inactive` branch earlier.

That switch is to skip a local Transceiver if it can't satisfy what the remote needs (so only sendrecv can work for sendrecv, sendonly and recvonly can satisify each other and inactive can't satisfy anything)